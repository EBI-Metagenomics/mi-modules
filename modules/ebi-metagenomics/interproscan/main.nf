process INTERPROSCAN {
    tag "$meta.id"
    label 'process_long'

    container 'microbiome-informatics/genomes-pipeline.ips:5.62-94.0'
    containerOptions {
        if (workflow.containerEngine == 'singularity') {
            return "--bind ${interproscan_db}/data:/opt/interproscan-5.62-94.0/data"
        } else {
            return "-v ${interproscan_db}/data:/opt/interproscan-5.62-94.0/data"
        }
    }

    input:
    tuple val(meta), path(fasta)
    tuple path(interproscan_db), val(db_version)
    val(out_ext)

    output:
    tuple val(meta), path('*.tsv') , optional: true, emit: tsv
    tuple val(meta), path('*.xml') , optional: true, emit: xml
    tuple val(meta), path('*.gff3'), optional: true, emit: gff3
    tuple val(meta), path('*.json'), optional: true, emit: json
    path "versions.yml"            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    def appl = "-appl TIGRFAM,SFLD,SUPERFAMILY,Gene3D,Hamap,Coils,ProSiteProfiles,SMART,CDD,PRINTS,PIRSF,ProSitePatterns,PfamA,ProDom,MobiDBLite,SignalP_GRAM_POSITIVE,SignalP_GRAM_NEGATIVE,SignalP_EUK,Phobius,TMHMM"
    if ( args.contains("-appl") ) {
        appl = ""
    }
    switch ( out_ext ) {
        case "tsv": break
        case "xml": break
        case "gff3": break
        case "json": break
        default:
            out_ext = 'tsv';
            log.warn("Unknown output file format provided (${out_ext}): selecting tsv as fallback");
            break
    }

    //  -dp (disable precalculation) is on so no online dependency
    """
    interproscan.sh \\
        -cpu $task.cpus \\
        -i $fasta \\
        -f ${out_ext} \\
        -dp \\
        ${appl} \\
        ${args} \\
        -o ${prefix}.${out_ext}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        InterProScan: \$(interproscan.sh --version | grep -o "InterProScan version [0-9.-]*" | sed "s/InterProScan version //")
        InterProScan database: $db_version
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    switch ( out_ext ) {
        case "tsv": break
        case "xml": break
        case "gff3": break
        case "json": break
        default:
            out_ext = 'tsv';
            log.warn("Unknown output file format provided (${out_ext}): selecting tsv as fallback");
            break
    }

    """
    touch ${prefix}.${out_ext}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        InterProScan: \$(interproscan.sh --version | grep -o "InterProScan version [0-9.-]*" | sed "s/InterProScan version //")
        InterProScan database: $db_version
    END_VERSIONS
    """
}
