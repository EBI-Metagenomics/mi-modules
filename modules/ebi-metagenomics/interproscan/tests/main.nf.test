// This krona/ktimporttext unit test file is converted to nf-test from an existing pytest test at this commit: https://github.com/nf-core/modules/commit/5e34754d42cd2d5d248ca8673c0a53cdf5624905
// I have converted it because the standard way of writing nf-core unit tests now is to use nf-test
// It is otherwise exactly the same unit test

nextflow_process {

    name "Test Process INTERPROSCAN"
    script "../main.nf"
    process "INTERPROSCAN"

    tag "modules"
    tag "modules_nfcore"
    tag "interproscan"

    test("test_interproscan") {

        when {
            process {
                """                
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/test.fasta", checkIfExists: true)
                    ]
                input[1] = [
                    file("${moduleDir}/tests/interproscan_db", type: 'dir', checkIfExists: true),
                    "5.62-94.0"
                ]
                input[2] = "tsv"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
