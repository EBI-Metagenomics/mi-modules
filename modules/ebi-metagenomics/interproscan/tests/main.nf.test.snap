{
    "test_interproscan": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,0a34cde772087a3023970ecd15299699"
                ],
                "gff3": [
                    
                ],
                "json": [
                    
                ],
                "tsv": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0a34cde772087a3023970ecd15299699"
                ],
                "xml": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-06-19T10:08:19.616771598"
    }
}